File: chunker/latex/formatLatex.ts
/**
 * Converts LaTeX expressions into human‑readable UTF‑8 text.
 * - $...$ / $$...$$ wrappers are stripped.
 * - Common commands are replaced with Unicode symbols.
 * - Subscripts / superscripts are kept as _{…} / ^{…}.
 * - Fractions become (a)/(b).
 */

export function formatLatex(text: string): string {
  let cleaned = text.replace(/\$\$([\s\S]*?)\$\$/g, "$1"); // display math
  cleaned = cleaned.replace(/\$([^$]+)\$/g, "$1");          // inline math

  cleaned = cleaned

    // ── Lowercase Greek ────────────────────────────────────────────
    .replace(/\\alpha\b/g, "α")
    .replace(/\\beta\b/g, "β")
    .replace(/\\gamma\b/g, "γ")
    .replace(/\\delta\b/g, "δ")
    .replace(/\\epsilon\b/g, "ε")
    .replace(/\\zeta\b/g, "ζ")
    .replace(/\\eta\b/g, "η")
    .replace(/\\theta\b/g, "θ")
    .replace(/\\iota\b/g, "ι")
    .replace(/\\kappa\b/g, "κ")
    .replace(/\\lambda\b/g, "λ")
    .replace(/\\mu\b/g, "μ")
    .replace(/\\nu\b/g, "ν")
    .replace(/\\xi\b/g, "ξ")
    .replace(/\\pi\b/g, "π")
    .replace(/\\rho\b/g, "ρ")
    .replace(/\\sigma\b/g, "σ")
    .replace(/\\tau\b/g, "τ")
    .replace(/\\upsilon\b/g, "υ")
    .replace(/\\phi\b/g, "φ")
    .replace(/\\chi\b/g, "χ")
    .replace(/\\psi\b/g, "ψ")
    .replace(/\\omega\b/g, "ω")

    // ── Uppercase Greek ────────────────────────────────────────────
    .replace(/\\Gamma\b/g, "Γ")
    .replace(/\\Delta\b/g, "Δ")
    .replace(/\\Theta\b/g, "Θ")
    .replace(/\\Lambda\b/g, "Λ")
    .replace(/\\Xi\b/g, "Ξ")
    .replace(/\\Pi\b/g, "Π")
    .replace(/\\Sigma\b/g, "Σ")
    .replace(/\\Upsilon\b/g, "Υ")
    .replace(/\\Phi\b/g, "Φ")
    .replace(/\\Psi\b/g, "Ψ")
    .replace(/\\Omega\b/g, "Ω")

    // ── Logic operators ────────────────────────────────────────────
    .replace(/\\forall(?![a-zA-Z])/g, "∀")       // universal quantifier
    .replace(/\\exists(?![a-zA-Z])/g, "∃")       // existential quantifier
    .replace(/\\neg(?![a-zA-Z])|\\lnot(?![a-zA-Z])/g, "¬")   // negation
    .replace(/\\land(?![a-zA-Z])|\\wedge(?![a-zA-Z])/g, "∧") // and
    .replace(/\\lor(?![a-zA-Z])|\\vee(?![a-zA-Z])/g, "∨")    // or
    .replace(/\\rightarrow(?![a-zA-Z])|\\to(?![a-zA-Z])/g, "→")       // implication
    .replace(/\\leftrightarrow(?![a-zA-Z])/g, "↔")                     // iff
    .replace(/\\Rightarrow(?![a-zA-Z])/g, "⇒")                         // double arrow
    .replace(/\\Leftrightarrow(?![a-zA-Z])/g, "⇔")                     // double iff
    .replace(/\\top(?![a-zA-Z])/g, "⊤")                                // top / true
    .replace(/\\bot(?![a-zA-Z])/g, "⊥")                                // bottom / false
    .replace(/\\vdash(?![a-zA-Z])/g, "⊢")                              // entails
    .replace(/\\models(?![a-zA-Z])/g, "⊨")                             // models
    .replace(/\\equiv(?![a-zA-Z])/g, "≡")                              // equivalence

    // ── Large operators (with subscript / superscript) ─────────────
    .replace(/\\sum\s*_\{([^}]+)\}\s*\^\{([^}]+)\}/g, "∑_{$1}^{$2}")
    .replace(/\\sum\s*_\{([^}]+)\}/g, "∑_{$1}")
    .replace(/\\sum\s*\^\{([^}]+)\}/g, "∑^{$1}")
    .replace(/\\prod\s*_\{([^}]+)\}\s*\^\{([^}]+)\}/g, "∏_{$1}^{$2}")
    .replace(/\\prod\s*_\{([^}]+)\}/g, "∏_{$1}")
    .replace(/\\prod\s*\^\{([^}]+)\}/g, "∏^{$1}")

    // ── Plain large operators (no limits) ──────────────────────────
    .replace(/\\sum(?![a-zA-Z])/g, "∑")
    .replace(/\\prod(?![a-zA-Z])/g, "∏")
    .replace(/\\int(?![a-zA-Z])/g, "∫")
    .replace(/\\infty(?![a-zA-Z])/g, "∞")

    // ── Arithmetic / binary operators ──────────────────────────────
    .replace(/\\cdot(?![a-zA-Z])/g, "·")
    .replace(/\\times(?![a-zA-Z])/g, "×")
    .replace(/\\div(?![a-zA-Z])/g, "÷")
    .replace(/\\pm(?![a-zA-Z])/g, "±")
    .replace(/\\mp(?![a-zA-Z])/g, "∓")
    .replace(/\\ast(?![a-zA-Z])/g, "*")

    // ── Relations ──────────────────────────────────────────────────
    .replace(/\\leq(?![a-zA-Z])/g, "≤")
    .replace(/\\geq(?![a-zA-Z])/g, "≥")
    .replace(/\\neq(?![a-zA-Z])/g, "≠")
    .replace(/\\approx(?![a-zA-Z])/g, "≈")
    .replace(/\\propto(?![a-zA-Z])/g, "∝")
    .replace(/\\sim(?![a-zA-Z])/g, "∼")

    // ── Arrows ─────────────────────────────────────────────────────
    .replace(/\\leftarrow(?![a-zA-Z])/g, "←")
    .replace(/\\Rightarrow(?![a-zA-Z])/g, "⇒")
    .replace(/\\Leftrightarrow(?![a-zA-Z])/g, "⇔")

    // ── Accents (vec, hat, tilde, bar) ─────────────────────────────
    .replace(/\\vec\{([^}]+)\}/g, "$1⃗")
    .replace(/\\hat\{([^}]+)\}/g, "$1̂")
    .replace(/\\tilde\{([^}]+)\}/g, "$1̃")
    .replace(/\\bar\{([^}]+)\}/g, "$1̅")

    // ── Fractions ──────────────────────────────────────────────────
    .replace(/\\frac\{([^}]+)\}\{([^}]+)\}/g, "($1)/($2)")

    // ── Final cleanup ──────────────────────────────────────────────
    .replace(/[{}]/g, "")              // remove stray braces
    .replace(/\\[a-zA-Z]+/g, "");       // drop remaining unknown commands

  return cleaned.trim();
}

File: src-tauri/src/chunker/latex/format_latex.rs
use regex::Regex;
use std::sync::LazyLock;
use std::borrow::Cow;

use std::collections::HashMap;

static SYMBOL_MAP: LazyLock<HashMap<&'static str, &'static str>> = LazyLock::new(|| {
    HashMap::from([
        ("alpha", "α"), ("beta", "β"), ("gamma", "γ"),
        ("delta", "δ"), ("epsilon", "ε"), ("zeta", "ζ"),
        ("eta", "η"), ("theta", "θ"), ("iota", "ι"),
        ("kappa", "κ"), ("lambda", "λ"), ("mu", "μ"),
        ("nu", "ν"), ("xi", "ξ"), ("pi", "π"),
        ("rho", "ρ"), ("sigma", "σ"), ("tau", "τ"),
        ("upsilon", "υ"), ("phi", "φ"), ("chi", "χ"),
        ("psi", "ψ"), ("omega", "ω"),
        
        ("Gamma", "Γ"), ("Delta", "Δ"), ("Theta", "Θ"),
        ("Lambda", "Λ"), ("Xi", "Ξ"), ("Pi", "Π"),
        ("Sigma", "Σ"), ("Upsilon", "Υ"), ("Phi", "Φ"),
        ("Psi", "Ψ"), ("Omega", "Ω"),

        ("forall", "∀"), ("exists", "∃"),
        ("neg", "¬"), ("lnot", "¬"), ("land", "∧"), ("wedge", "∧"),
        ("lor", "∨"), ("vee", "∨"), ("rightarrow", "→"), ("to", "→"),
        ("leftrightarrow", "↔"), ("Rightarrow", "⇒"),
        ("Leftrightarrow", "⇔"), ("top", "⊤"),
        ("bot", "⊥"), ("vdash", "⊢"),
        ("models", "⊨"), ("equiv", "≡"),

        ("sum", "∑"), ("prod", "∏"),
        ("int", "∫"), ("infty", "∞"),

        ("cdot", "·"), ("times", "×"),
        ("div", "÷"), ("pm", "±"),
        ("mp", "∓"), ("ast", "*"),

        ("leq", "≤"), ("geq", "≥"),
        ("neq", "≠"), ("approx", "≈"),
        ("propto", "∝"), ("sim", "∼"),

        ("leftarrow", "←"),
    ])
});

static RE_LATEX_SYMBOLS: LazyLock<Regex> = LazyLock::new(|| {
    let keys: Vec<&str> = SYMBOL_MAP.keys().copied().collect();
    let pattern = format!(r"\\({})(?-u:\b)", keys.join("|"));
    Regex::new(&pattern).unwrap()
});

static RE_LATEX_BLOCK: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"(?s)\$\$([\s\S]*?)\$\$").unwrap());
static RE_LATEX_INLINE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"(?s)\$([^$]+)\$").unwrap());

static RE_SUM_SUB_SUP: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\\sum\s*_\{([^}]+)\}\s*\^\{([^}]+)\}").unwrap());
static RE_SUM_SUB: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\\sum\s*_\{([^}]+)\}").unwrap());
static RE_SUM_SUP: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\\sum\s*\^\{([^}]+)\}").unwrap());
static RE_PROD_SUB_SUP: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\\prod\s*_\{([^}]+)\}\s*\^\{([^}]+)\}").unwrap());
static RE_PROD_SUB: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\\prod\s*_\{([^}]+)\}").unwrap());
static RE_PROD_SUP: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\\prod\s*\^\{([^}]+)\}").unwrap());
static RE_VEC: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\\vec\{([^}]+)\}").unwrap());
static RE_HAT: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\\hat\{([^}]+)\}").unwrap());
static RE_TILDE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\\tilde\{([^}]+)\}").unwrap());
static RE_BAR: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\\bar\{([^}]+)\}").unwrap());
static RE_FRAC: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\\frac\{([^}]+)\}\{([^}]+)\}").unwrap());
static RE_BRACES: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"[{}]").unwrap());
static RE_UNKNOWN: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\\[a-zA-Z]+").unwrap());

pub fn format_latex(text: &str) -> Cow<'_, str> {
    let text1 = RE_LATEX_BLOCK.replace_all(text, |caps: &regex::Captures| {
        format!("$${}$$", transform_math(&caps[1]))
    });
    
    match RE_LATEX_INLINE.replace_all(&text1, |caps: &regex::Captures| {
        format!("${}$", transform_math(&caps[1]))
    }) {
        Cow::Owned(s) => Cow::Owned(s),
        Cow::Borrowed(_) => text1,
    }
}

fn transform_math(s: &str) -> String {
    let mut cleaned = Cow::Borrowed(s);

    if let Cow::Owned(new_s) = RE_LATEX_SYMBOLS.replace_all(&cleaned, |caps: &regex::Captures| {
        SYMBOL_MAP.get(&caps[1]).unwrap().to_string()
    }) {
        cleaned = Cow::Owned(new_s);
    }

    if let Cow::Owned(new_s) = RE_SUM_SUB_SUP.replace_all(&cleaned, "∑_{$1}^{$2}") { cleaned = Cow::Owned(new_s); }
    if let Cow::Owned(new_s) = RE_SUM_SUB.replace_all(&cleaned, "∑_{$1}") { cleaned = Cow::Owned(new_s); }
    if let Cow::Owned(new_s) = RE_SUM_SUP.replace_all(&cleaned, "∑^{$1}") { cleaned = Cow::Owned(new_s); }
    
    if let Cow::Owned(new_s) = RE_PROD_SUB_SUP.replace_all(&cleaned, "∏_{$1}^{$2}") { cleaned = Cow::Owned(new_s); }
    if let Cow::Owned(new_s) = RE_PROD_SUB.replace_all(&cleaned, "∏_{$1}") { cleaned = Cow::Owned(new_s); }
    if let Cow::Owned(new_s) = RE_PROD_SUP.replace_all(&cleaned, "∏^{$1}") { cleaned = Cow::Owned(new_s); }

    if let Cow::Owned(new_s) = RE_VEC.replace_all(&cleaned, "$1⃗") { cleaned = Cow::Owned(new_s); }
    if let Cow::Owned(new_s) = RE_HAT.replace_all(&cleaned, "$1̂") { cleaned = Cow::Owned(new_s); }
    if let Cow::Owned(new_s) = RE_TILDE.replace_all(&cleaned, "$1̃") { cleaned = Cow::Owned(new_s); }
    if let Cow::Owned(new_s) = RE_BAR.replace_all(&cleaned, "$1̅") { cleaned = Cow::Owned(new_s); }
    if let Cow::Owned(new_s) = RE_FRAC.replace_all(&cleaned, "($1)/($2)") { cleaned = Cow::Owned(new_s); }
    if let Cow::Owned(new_s) = RE_BRACES.replace_all(&cleaned, "") { cleaned = Cow::Owned(new_s); }
    if let Cow::Owned(new_s) = RE_UNKNOWN.replace_all(&cleaned, "") { cleaned = Cow::Owned(new_s); }

    cleaned.into_owned()
}
